Server/App/utils/database.py

```diff
@@ -25,22 +25,33 @@ def __del__(self):
     def begin_transaction(self):
         """Start a transaction by disabling autocommit."""
         self.connector.autocommit(False)
+        self.num_nested_transactions += 1
 
     def end_transaction(self):
         """Commit and then reset the autocommit to True."""
+        self.num_nested_transactions -= 1
+        if self.num_nested_transactions > 0:
+            return
         self.connector.commit()
         self.connector.autocommit(True)
 
     def abort_transaction(self):
         """Abort the transaction and reset the autocommit to True."""
+        self.num_nested_transactions -= 1
+        if self.num_nested_transactions > 0:
+            return
         self.connector.rollback()
         self.connector.autocommit(True)
 
     def call_procedure_transactionally(self, proc_name, *params):
         """Call a stored procedure in the database and JSON-ify the results."""
-        self.begin_transaction()
-        result = self.call_procedure(proc_name, *params)
-        self.end_transaction()
+        try:
+            self.begin_transaction()
```

Comment:
Yeah I tend to wrap the procedures that write in a transaction. If the procedure is read only then I don't usually wrap it in a transaction. Also, if there are multiple procedures that need to get called to modify something, I wrap them in one bigger transaction.